### PR TITLE
fix: cache network calls during a single run

### PR DESCRIPTION
**Pull request by mchandak29**
implements the idea in issue https://github.com/adobe/aemscreens-offlineresources-generator/issues/8

When tested for CEC(with 26 channels) - reduces the network calls from 550 to just 65.

### CODE REVIEW

**Comment by habansal** [comment]
@mchandak29 what is the performance benefit (total workflow running time) for the same dataset?

File: src/utils/fetchUtils.js [hunk 1]
```diff
@@ -25,6 +27,9 @@ export default class FetchUtils {
   };
 
   static fetchDataFromUrl = async (url, additionalHeaders = {}) => {
+    if (global.cache[url]) {
+      return Promise.resolve(global.cache[url]);
```
Comment (habansal) [inline src/utils/fetchUtils.js:31]: Are we sending resolved Promises from the cache?
In this same method, we are sending an unresolved promise -
```
 global.cache[url] = response.text();
return global.cache[url];
```

File: src/utils/fetchUtils.js [hunk 2]
```diff
@@ -37,11 +42,33 @@ export default class FetchUtils {
           if (!response.ok) {
             throw new Error(`request to fetch ${url} failed with status code ${response.status}`);
           }
-          return response.text();
+          global.cache[url] = response.text();
```
Comment (habansal) [inline src/utils/fetchUtils.js:45]: In the `makeHeadRequest` method, we are storing response objects from HEAD calls directly under cache and here, we are storing response.text().
Should we add nesting e.g. somethingLike -
`global.cache.response, global.cache.responseText` maybe with better names (if needed)

**Comment by mchandak29** [comment]
> what is the performance benefit (total workflow running time) for the same dataset (26 channels)?

time to generate offline resources went from **59s** to **7s**.
total time went from **1m32s** to **47s**

before - https://github.com/hlxscreens/cec-screens/actions/runs/5670587447
after - https://github.com/hlxscreens/cec-screens/actions/runs/5678311557

File: src/utils/fetchUtils.js (hunk 1 above)
Comment (mchandak29) [reply src/utils/fetchUtils.js:31]: we are sending promise in both cases. see L#48
both of which resolves to `global.cache[url]`

but let me go ahead and improve readability of this code.

Comment (habansal) [reply src/utils/fetchUtils.js:31]: For the sake of clarification, On line#31, can't we send -
```global.cache[url]```
Isn't using `Promise.resolve` redundant here?

File: src/createManifest.js [hunk 3]
```diff
@@ -81,27 +80,25 @@ export default class ManifestGenerator {
...
+        if (!(await GitUtils.isFileDirty(resourceSubPath.slice(1)))) {
+          console.log(`resource ${resourceSubPath} not available for channel ${path}`);
+          // eslint-disable-next-line no-continue
+          continue;
+        }
       }
       const resourceEntry = {};
       resourceEntry.path = resourcesArr[i];
       // timestamp is optional value, only add if last-modified available
-      const date = resp.headers.get('last-modified');
```
Comment (habansal) [inline src/createManifest.js:100]: [Minor] Could this work instead -

```
const date = resp && resp.headers.get('last-modified');
} else if (date) {
```

File: src/utils/fetchUtils.js [hunk 4]
```diff
@@ -19,29 +19,35 @@ export default class FetchUtils {
...
-          if (!response.ok) {
-            throw new Error(`request to fetch ${url} failed with status code ${response.status}`);
-          }
-          return response.text();
-        });
-      return Promise.resolve(result);
+      });
+
+      if (!response.ok) {
+        throw new Error(`Request to fetch ${url} failed with status code ${response.status}`);
```
Comment (habansal) [inline src/utils/fetchUtils.js:45]: will the response be cached in this case as well ?

File: src/utils/fetchUtils.js [hunk 5]
```diff
@@ -37,11 +42,33 @@ export default class FetchUtils {
...
+        resourcePath,
+        {
+          method: 'HEAD',
+          headers: {
+            'x-franklin-allowlist-key': process.env.franklinAllowlistKey,
+            ...additionalHeaders
+          }
+        }
+      );
+      global.cache[resourcePath] = resp;
```
[the remark this answers is not in the document]
Comment (habansal) [reply src/utils/fetchUtils.js]: Fixed. Resolving the comment

File: src/utils/fetchUtils.js (hunk 1 above)
Comment (habansal) [reply src/utils/fetchUtils.js]: Fixed with the refactor to use `node-fetch-cache`. Resolving the comment

File: src/utils/fetchUtils.js (hunk 2 above)
Comment (habansal) [reply src/utils/fetchUtils.js]: Fixed via use of `node-fetch-cache`. Resolving the comment

Comment (habansal) [reply src/utils/fetchUtils.js]: Fixed via use of `node-fetch-cache`. Resolving the comment.

File: src/createManifest.js [hunk 6]
```diff
@@ -51,7 +50,7 @@ export default class ManifestGenerator {
    */
   static getPageJsonEntry = async (host, path, isHtmlUpdated) => {
     const entryPath = `${path}.html`;
-    const resp = await FetchUtils.fetchData(host, path);
+    const resp = await FetchUtils.fetchDataWithMethod(host, path, 'HEAD');
```
[the remark this answers is not in the document]
Comment (mchandak29) [reply src/createManifest.js:53]: yes, that is expected

File: src/utils/fetchUtils.js (hunk 4 above)
Comment (mchandak29) [reply src/utils/fetchUtils.js:47]: removed the error response from cache

File: src/createManifest.js [hunk 7]
```diff
@@ -81,27 +80,25 @@ export default class ManifestGenerator {
...
       }
       const resourceEntry = {};
       resourceEntry.path = resourcesArr[i];
       // timestamp is optional value, only add if last-modified available
-      const date = resp.headers.get('last-modified');
       if (ManifestGenerator.isMedia(resourceSubPath)) {
         resourceEntry.path = parentPath.concat(resourceEntry.path);
         resourceEntry.hash = ManifestGenerator.getHashFromMedia(resourceSubPath);
-      } else if (date) {
+      } else if (resp && resp.headers.get('last-modified')) {
```
[the remark this answers is not in the document]
Comment (habansal) [reply src/createManifest.js]: Fixed. Resolving comment

File: src/utils/fetchUtils.js (hunk 4 above)
Comment (habansal) [reply src/utils/fetchUtils.js:47]: Looks good.

File: src/createManifest.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (habansal) [reply src/createManifest.js:53]: Cool, Mentioning in JsDoc would help

**Review by habansal — DISMISSED** [review]

**Pull request merged**